src/Gaffer/MetadataAlgo.cpp

```diff
@@ -222,6 +224,89 @@ void bookmarks( const Node *node, std::vector<NodePtr> &bookmarks )
 	}
 }
 
+void setNumericBookmark( Node *node, int value )
+{
+	if( value < 0 )
+	{
+		return;
+	}
+
+	if( readOnly( node ) )
+	{
+		return;
+	}
+
+	// \todo: should this be const (and the `root` arg in nodesWithMetadata as well?)
+	ScriptNode *scriptNode = node->ancestor<ScriptNode>();
+	if( !scriptNode )
+	{
+		return;
+	}
+
+	for( Node *nodeWithMetadata : Metadata::nodesWithMetadata( scriptNode, g_numericBookmarkName, /* instanceOnly = */ true ) )
+	{
+		if( getNumericBookmark( nodeWithMetadata ) != value )
+		{
+			continue;
+		}
+
+		if( readOnly( nodeWithMetadata ) )
+		{
+			continue;
+		}
+
+		Metadata::deregisterValue( nodeWithMetadata, g_numericBookmarkName );
+	}
+
+	Metadata::registerValue( node, g_numericBookmarkName, new IntData( value ), false );
+}
+
+int getNumericBookmark( const Node *node )
+{
+	ConstIntDataPtr i = Metadata::value<IntData>( node, g_numericBookmarkName );
+	if( !i || readOnly( node ) )
+	{
+		return -1;
+	}
+
+	return i->readable();
+}
+
+void removeNumericBookmark( Node *node )
+{
+	if( readOnly( node ) )
+	{
+		return;
+	}
+
+	Metadata::deregisterValue( node, g_numericBookmarkName );
+}
+
+// \todo: const parent? same for nodesWithMetadata?
```

Comment:
Yes, definitely. Let's get this right in the header and use a const_cast in here, and then remove the const_cast in the same non-backportable commit I mentioned above. Although actually, I'm not sure this method would be necessary if we had the API I suggest above - do we ever want to query all bookmarks at once?